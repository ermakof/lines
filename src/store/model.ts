import IUserProfile from '@src/modules/Auth/model/IUserProfile';

export interface IState {
  isLoading?: boolean;
  gameFieldSize: number;
  gameFieldPercentFilled: number;
  gameFieldData: Array<number>;
  selectedCell?: number;
  userLevel: string;
  userProfile?: IUserProfile;
}
