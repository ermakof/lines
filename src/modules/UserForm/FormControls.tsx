/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo } from 'react';
import { useDispatch } from 'react-redux';

import Button from '@src/components/Button';
import { actions as appActions } from '@src/App/appSlice';
import { actions as authActions } from '@src/modules/Auth/authSlice';
import Select from '@src/components/Select';
import styled from '@emotion/styled';

const Root = styled.div`
  display: flex;
`;

interface IFormControls {
  userLevel?: string;
}
const FormControls: FC<IFormControls> = ({ userLevel }) => {
  const dispatch = useDispatch();

  const handleReset = () => {
    dispatch(appActions.resetApp());
  };

  const handleSelectLevel = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { value } = event.currentTarget;
    dispatch(appActions.setUserLevel(value));
  };

  const handleLogout = () => {
    dispatch(authActions.logout());
  };

  return (
    <Root>
      <Button role="buttonReset" onClick={handleReset} title="Сброс" />
      <Select onSelect={handleSelectLevel} selectedLevel={userLevel} />
      <Button role="buttonLogout" onClick={handleLogout} title="Выйти" />
    </Root>
  );
};

export default memo(FormControls);
