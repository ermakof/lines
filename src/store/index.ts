import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';

import rootReducer from '@src/store/rootReducer';

const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware()
      .prepend
      // correctly typed middlewares can just be used
      ()
      // prepend and concat calls can be chained
      .concat(logger),
});

type TRootState = ReturnType<typeof store.getState>;

export { store, TRootState };
