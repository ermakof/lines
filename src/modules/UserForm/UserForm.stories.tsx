import React from 'react';
import { Provider } from 'react-redux';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';
import { BrowserRouter as Router } from 'react-router-dom';

import UserForm from '@src/modules/UserForm';
import { store } from '@src/store';

const dispatch = () => null;

const state = {
  gameLevel: '1',
  gameFieldSize: 4,
  gameFieldPercentFilled: 10,
  gameFieldData: [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
};

export default {
  component: UserForm,
  decorators: [withKnobs],
  title: 'Forms/UserForm',
} as ComponentMeta<typeof UserForm>;

const Template: ComponentStory<typeof UserForm> = (args) => (
  <Provider store={store}>
    <Router>
      <UserForm {...args} />
    </Router>
  </Provider>
);
export const Static = Template.bind({});

Static.args = {
  onSubmit: action('submitted'),
};
