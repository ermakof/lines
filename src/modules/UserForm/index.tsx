/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Panel from '@src/layout/Panel';
import { actions as appActions } from '@src/App/appSlice';
import { TRootState } from '@src/store';
import FormControls from '@src/modules/UserForm/FormControls';

const UserForm: FC = () => {
  const { userLevel } = useSelector(({ app }: TRootState) => app);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(appActions.rehydrate());
  }, []);

  return (
    <Panel role="userForm">
      <FormControls userLevel={userLevel} />
    </Panel>
  );
};

export default memo(UserForm);
