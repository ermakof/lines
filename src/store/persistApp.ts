import { LOCAL_STORAGE_APP_KEY, store } from '@src/store/index';

const persistApp = () => {
  const app = store.getState().app;
  localStorage.setItem(LOCAL_STORAGE_APP_KEY, JSON.stringify(app));
};

export default persistApp;
