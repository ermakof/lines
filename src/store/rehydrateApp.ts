import { actions } from '@src/modules/Auth/authSlice';
import { store } from '@src/store/index';

export const rehydrateApp = () => {
  store.dispatch(actions.rehydrate());
};
