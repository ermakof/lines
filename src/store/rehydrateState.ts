import appSlice from '@src/App/appSlice';
import { LOCAL_STORAGE_APP_KEY, store } from '@src/store/index';

const rehydrateState = () => {
  try {
    const persistedState = localStorage.getItem(LOCAL_STORAGE_APP_KEY);
    if (persistedState) {
      const payload = JSON.parse(persistedState);
      store.dispatch(appSlice.actions.hydrate(payload));
    }
  } catch (e) {
    console.log(e);
  }
};

export default rehydrateState;
