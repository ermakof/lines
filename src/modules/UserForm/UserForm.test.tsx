import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

import UserForm from '@src/modules/UserForm';
import { store } from '@src/store';

describe('UserForm', () => {
  it('Render <UserForm>', () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    };
    const dispatch = jest.fn();
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const userForm = screen.getByRole(/userForm/gi);
    expect(userForm).toBeInTheDocument();
    const buttonReset = screen.getByRole(/^buttonReset$/gi);
    expect(buttonReset).toBeInTheDocument();
    const select = screen.getByRole(/^select$/gi);
    expect(select).toBeInTheDocument();
    const buttonLogout = screen.getByRole(/^buttonLogout$/gi);
    expect(buttonLogout).toBeInTheDocument();
  });

  it('Reset game <UserForm>', async () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    };
    const dispatch = jest.fn();
    render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    userEvent.click(screen.getByRole('buttonReset'));

    await waitFor(() =>
      expect(dispatch).toHaveBeenCalledWith({
        type: 'APP__RESET',
      })
    );
  });

  it('Logout app <UserForm>', async () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    };
    const dispatch = jest.fn();
    render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    userEvent.click(screen.getByRole('buttonLogout'));

    await waitFor(() =>
      expect(dispatch).toHaveBeenCalledWith({
        type: 'APP__LOGOUT',
      })
    );
  });

  it('Select level <UserForm>', async () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    };
    const dispatch = jest.fn();
    render(
      <Provider store={store}>
        <Router>
          <UserForm />
        </Router>
      </Provider>
    );
    const select = screen.getByRole(/select/gi);
    userEvent.selectOptions(select, '2');
    const option = screen.getByRole('option', { name: 'Джедай' }) as HTMLOptionElement;
    expect(option.selected).toBe(true);

    await waitFor(() =>
      expect(dispatch).toHaveBeenCalledWith({
        payload: { gameLevel: '2' },
        type: 'APP__SET_LEVEL',
      })
    );
  });
});
