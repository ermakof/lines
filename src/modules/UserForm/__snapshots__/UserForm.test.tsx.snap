// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Render <UserForm> 1`] = `
<DocumentFragment>
  <div
    class="css-1w0hfqs"
    role="userForm"
  >
    <div
      class="css-ov1ktg"
    >
      <button
        class="css-tv5zn1"
        role="buttonReset"
      >
        Сброс
      </button>
      <div
        class="css-1xvr7bb"
      >
        <p>
          Уровень
        </p>
        <select
          name="select"
          role="select"
        >
          <option
            role="option"
            value="1"
          >
            Падаван
          </option>
          <option
            role="option"
            value="2"
          >
            Джедай
          </option>
          <option
            role="option"
            value="3"
          >
            Член Совета Силы
          </option>
        </select>
      </div>
      <button
        class="css-tv5zn1"
        role="buttonLogout"
      >
        Выйти
      </button>
    </div>
  </div>
</DocumentFragment>
`;
